### DISCUSSION TITLE
Extending the "named elements" feature of tuples to other types

### DISCUSSION THREAD

**Discussion by Richiban** [dotnet/csharplang#1883] [Language Ideas]
Tuples have this cool feature where the elements of the tuple (Item1, Item2 etc) can be named at compile time without having to declare a new type. These names are implemented via metadata attributes that appear on methods that accept / return tuple types, something like this:

```csharp
public (T1 value1, T2 value2) M();
```

which compiles to:

```csharp
[return: TupleElementNamesAttribute("value1", "value2")]
public ValueTuple<T1, T2> M();
```

Can we generalise this feature a bit more so that, at the point of instantiation, we can give more locally meaningful names to parameters, properties etc that might otherwise have completely generic labels?

Imagine that we have a dictionary that is enumerable as a sequence of tuples (instead of the current KeyValuePair<,>):

```csharp

public class MyDictionary<TKey, TValue>
{
	// ...

	public IEnumerator<(TKey, TValue)> GetEnumerator()
	{
		// ... 
	}

	// ...
}

```

I'd like to be able to instantiate it like this:

```csharp
var myDictionary = new MyDictionary<string english, string french>();
```

Then, when you enumerate it, you get a tuple with the names that were provided in the instantiation:

```csharp

foreach(var kv in myDictionary)
{
	Console.WriteLine($"{kv.english} => {kv.french}");
}

```

It would also be extremely useful for higher order functions, where `Func<int, int, int>` appears to have the signature `int arg1, int arg2`. This leads to one of the few occasions where Intellisense can't help you figure out how to use a method: e.g. the `Select` overload on `IEnumerable<T>`: `Select<T, R>(Func<T, int, R>)`. If, in this example, `T` is also `int` then I _always_ have to look up the docs to figure out which argument to the function is the element itself and which is the index. If the method signature could be written as:

```csharp
public IEnumerable<T> Select<T, R>(Func<T element, int index, R result> selector)
{
	// ...
}
```

then all ambiguity is removed.

In terms of implementation, I see no reason why this couldn't also be implemented through the use of attributes, although I must admit that my understanding of the underlying implementation if Tuples isn't very deep.

Thoughts welcome -- this is a relatively quickly thought-through feature.

**Comment by jnm2** [comment]
The same has been asked for Func and Action, and I would certainly like it for custom types like OneOf.

**Comment by HaloFour** [comment]
The issue I have with a feature like this is how awkward it would apply to most types.  With `ValueTuple<...>` and `Func<...>` you get a guarantee that each generic type parameter lines up with exactly a single argument or field.  But that's an exception and not a rule.  In the majority of cases a feature like this either wouldn't work at all or would result in really confusing behavior.

**Comment by svick** [comment]
> The same has been asked for Func and Action

That's https://github.com/dotnet/csharplang/issues/269.

**Comment by TheUnlocked** [comment]
@HaloFour One would expect that you could only do this for generics that have been specified with an attribute (or something similar). Like in most cases, if something doesn't make sense in a particular context, it shouldn't be used in that context.

**Comment by truegoodwill** [comment]
Please consider not just parameter naming but also documenation. It would be great to comment a Func field declaration like this: 

```csharp
/// <summary>My function</summary>
/// <param name="a">Description of parameter</param>
/// <param name="b">Description of parameter</param>
public Func<int a, int b, int> MyFunc { get; set; }
```

**Comment by SlimeNull** [comment]
Named tuple is really good idea! I love it!

**Comment by hamarb123** [comment]
I, too would really like this. I recently wrote custom `SequentialValueTuple` structs (essentially same as `ValueTuple`, but with `LayoutKind.Sequential` + a quick (essentially copy/nop most of the time) convertor) for use with pointers and interop and I'm really missing having the ability to name the fields.

> The issue I have with a feature like this is how awkward it would apply to most types. With ValueTuple<...> and Func<...> you get a guarantee that each generic type parameter lines up with exactly a single argument or field. But that's an exception and not a rule. In the majority of cases a feature like this either wouldn't work at all or would result in really confusing behavior.

@HaloFour, it could be done with an attribute like `[TupleGenericElementPosition(<int>)]` for non-delegate types, and the compiler could scan for which field/property/etc. has it, and use some kind of decision making for when it has multiple things with the same number (eg. 2 fields with `[TupleGenericElementPosition(3)]`) if the programmer was bad (should be an error in C#).

Edit: this could also be used with the .NET `Tuple` type, and perhaps `TupleGenericElementPosition` could be used on other parts of the type such as methods, and events.

**Comment by TobiasKnauss** [comment]
The suggestion for named keys and values was also made in #894

**Comment by najak3d** [comment]
This is an immensely useful feature, as it adds tremendous/easy clarity to most usages of Generics (some contexts much more than others).

**Very Useful Notation Examples:**
```c#
public Dictionary<string EnglishWord, string FrenchWord> EnglishToFrenchTranslations; 

public event Action<string name, int age>  BirthdayOccurred; 

private Func<string name, Date birthdate, bool isMarried, Person person> _CreatePerson;

```
**Less Useful, but still Helpful Notation:  (shows up in Intellisense)**
```c#
public class MyEntity<T EntityCategory> { }  
```

I'm surprised that this form of enhancement hasn't already made it's way into the C# language.   I guess it's just a matter of too big of a workload for too few people doing the work.

**Reply by CyrusNajmabadi** [reply]
We already have delegates for this case though. That's one it the reasons to use them.

**Reply by najak3d** [reply]
Your argument here would also argue to never use Tuples (anonymous records).    Anonymous Action/Func are a great idea - but if you want to use them, the current deficient notation makes them inferior to delegates, without good reason.

**In comparison, Delegates have these issues:**
1. Requires more work.  Now you have to think up a descriptive delegate Type name, and figure out where to put it in the API, and think "who else might use this?" and then attempt to make it adaptable for various other scenarios, so that you don't have to create too many of these delegates (as more types clutters the namespaces).
2. Now you have to advertise these delegate for others to use, else they'll create their own equivalent delegates, simply for not knowing that they already exists.
3. Intellisense/API - gets clutters, with lots more types now, which many only be used ONCE.
4. When someone sees this delegate on an API, they can't tell from looking at it, what are the parameters - if they want to know this, they have to "Go to Definition", which is extra work for the consumers of strong-type named delegates.

**Now Compare that to Anonymous methods (Action<T, T2, T3>):**
1. They are locally defined and fully explanatory.  Perfect for one-use cases.
2. You don't have to F12 to the definition to find out it's parameters.  It's all there in one spot.
3. You don't clutter up API with new Delegate Types (many only used once, or a few times)
4. You don't have to think about "re-use" and "where to put this new type" - for sake of re-use (only to reduce API members/clutter)

PLUS - this proposal also covers non-delegates, such as the "Dictionary" where it's also useful for replacing "Key/Value" with more meaningful accessors.

==
I'm not sure why you are pushing back on this?   This proposal has the same benefits as Tuples with named parameters, for the same reason.    After Tuples without parameter naming were invented.... **did you also argue against adding parameter names to Tuples??** Did you argue against it saying "but we already have classes and structs for this case though; we don't need Tuples". ???

Anonymous delegates were invented for good reason, and are in rampant use.  There's no good reason for handicapping them, when they easily could be improved in this vital way.

**Reply by HaloFour** [reply]
The names you'd apply to tuples are directly referencable in code.  Names you could apply to anonymous delegates can never be as the argument names are always redefined at the point of consumption.  As such they have very little value.

**Reply by najak3d** [reply]
For tuples, you could still access members as "Item 1/2/3"...   And so the work for making Tuple parameter naming was *more* work than it would be for Anonymous delegate parameter names.   For anonymous delegates - it's simply adding attributes that inform those consuming the delegates the meaning of parameter names -- and the auto-generated methods will have the correct/meaningful names.

This proposal also addresses things like Dictionary<K, T> -- which would behave like Tuples parameter naming, in that it would change the compiler parsing to allow for the named fields of the Dictionary types.

**Reply by RikkiGibson** [reply]
IMO, parameter names on anonymous delegates would be very useful for the "lambda authoring" scenario:

```cs
void M(string Name, int Id, Func<int, string, bool, Widget> factory) { }
void M(string Name, int Id, Widget (int Id, string Name, bool caseSensitive) factory) { }

// now you're calling the method
M("Halo", 4, ($$ // what do I type here? '(id, name, caseSensitive) =>' seems like a good start. But how would I know that with the first version of the method signature?
```

Relates to dotnet/roslyn#46190

**Reply by CyrusNajmabadi** [reply]
>  Now you have to think up a descriptive delegate Type name, 

Right. And that's the state of mind you're already in and the problem you're solving here. Giving descriptive names to things. So it's not unreasonable to continue that here.

**Reply by CyrusNajmabadi** [reply]
> Your argument here would also argue to never use Tuples (anonymous records). 

And I make that argument much of the time. I think tuples should be used sparingly and judiciously. Once you start needing more functionality, take the step to a record struct.

**Reply by CyrusNajmabadi** [reply]
> I'm not sure why you are pushing back on this? 

My reasons are listed. But the core thing to understand is that this is subjective, and people disagree with each other next because we vote these things differently.

**Reply by alrz** [reply]
This could also address the existing naming inconsistencies https://twitter.com/controlflow/status/1627634622069723136
```cs
var f = (int x) => { };
f(x: 42);

var g = (int x = -1) => x;
g(x: 42);
```
>` Widget (int Id, string Name, bool caseSensitive) factory`

There's already `delegate<...>` for that but it looks a little out of place.
